### PR TITLE
[Gradient Compression] Error feedback for PowerSGD (still need to fix the key in error_dict)

### PR DESCRIPTION
**Pull request by wayi1**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#48670 [Gradient Compression] Error feedback for PowerSGD (still need to fix the key in error_dict)**

Support an optional error feedback for PowerSGD -- storing the difference (i.e., the local error caused by compression) between the input gradient (adjusted by the existing error) and the gradient after decompression, and reinserting it at the next iteration.

Still need to add an index field to GradBucket as the key of error_dict. This is because the current key, input tensor of the bucket, can change across steps, as the buckets may be rebuilt in forward pass in order to save peak memory usage.

This is halfway of error feedback. Plan to add the new index field in a separate PR.

Original PR issue: Investigate Applying PowerSGD to Communication Hook for Gradient Compression #47202

Differential Revision: [D25240290](https://our.internmc.facebook.com/intern/diff/D25240290/)

### CODE REVIEW

File: torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py [hunk 1]
```diff
@@ -30,17 +30,44 @@ def _orthogonalize(matrix, epsilon=1e-8):
...
+        use_error_feedback=True,
+        random_seed=0,
+    ):
         self.process_group = process_group
         self.matrix_approximation_rank = matrix_approximation_rank
+        # Error feedback is usually crucial for both for convergence and generalization,
+        # because PowerSGD is a biased compressor,
+        # i.e., compressing and decompressing a random gradient does not yield the original in expectation.
+        # This mechanism requires a temporary copy of the input gradients,
+        # so it increases the peak memory consumption.
```
Comment (rohan-varma) [inline torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py:54]: I'm assuming that the increase in peak memory consumption is size of the gradient tensor? Should we clearly state this in the comment?

File: torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py [hunk 2]
```diff
@@ -30,17 +30,44 @@ def _orthogonalize(matrix, epsilon=1e-8):
...
+        # However, if the target matrices are known to be exactly low-ranked (instead of just low stable rank),
```
Comment (rohan-varma) [inline torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py:55]: When does this actually happen (target matrices are low-rank)?

File: torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py [hunk 3]
```diff
@@ -98,6 +125,16 @@ def powerSGD_hook(
     padded_total_length = square_side_length ** 2
     input_tensor.resize_(padded_total_length)
     input_tensor[total_length:padded_total_length].fill_(0)
+
+    # Incorporate the error from the previous state into the gradients.
+    if state.use_error_feedback:
+        if input_tensor not in state.error_dict:
+            state.error_dict[input_tensor] = torch.zeros(padded_total_length, device=device)
+        input_tensor.add_(state.error_dict[input_tensor])
```
Comment (rohan-varma) [inline torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py:133]: I guess we're adding zeros if the above if statement is true, should we bypass unnecessarily adding zeros in this case for performance?

File: torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py [hunk 4]
```diff
@@ -141,6 +178,9 @@ def decompress(fut):
         q = fut.value()[0].div_(world_size)
         torch.matmul(p, q.t(), out=matrix)
 
+        if state.use_error_feedback:
+            # Memorize the local errors.
+            state.error_dict[input_tensor] = input_tensor_cp - input_tensor
```
Comment (rohan-varma) [inline torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py:183]: As I understand this algorithm, iteration i + 1 needs iteration i's `input_tensor`, to save memory usage, can we clean up the unneeded tensors from the dict from previous states and would this improve the memory consumption situation?

Comment (rohan-varma) [inline torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py:183]: I'm not completely familiar with the differeent types of errors tha we can look at, but would we be interested in different p-norms instead of just l1 difference? Also, would we be interested in keeping a sort of decaying weighted average (similar to momentum) of errors instead of the error being 100% decided by the previous iteration difference?

**Review by rohan-varma — COMMENTED** [review]
Looks good overall, have a couple of comments inline.

File: torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py [hunk 5]
```diff
@@ -30,17 +30,44 @@ def _orthogonalize(matrix, epsilon=1e-8):
...
+        "rng",
+        "error_dict",
+    ]
+
+    def __init__(
+        self,
+        process_group,
+        matrix_approximation_rank=1,
+        use_error_feedback=True,
+        random_seed=0,
```
Comment (rohan-varma) [inline torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py:46]: Once we formalize the API around powerSGDState, would be great if we could add concrete comuentation for each of these parameters and the tradeoffs involved in using/not using them

**Review by rohan-varma — APPROVED** [review]

File: torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py (hunk 5 above)
Comment (rohan-varma) [reply torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py:46]: Let's also make sure to add unit tests either in this Pr or follow up PRs.

File: torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py (hunk 1 above)
Comment (SciPioneer) [reply torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py:54]: I think it should be clear, since I mentioned "This mechanism requires a temporary copy of the input gradients,"

To make it more explicitly:
"so it increases the peak memory consumption."
->
"so it increases the peak memory consumption by the size of gradient tensor."

File: torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py (hunk 3 above)
Comment (SciPioneer) [reply torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py:133]: Thanks! Add `else` to avoid the unnecessary execution, although adding zeros only occur once at the first iteration.

File: torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py (hunk 4 above)
Comment (SciPioneer) [reply torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py:183]: Thanks for the interesting thoughts! For now, let's first follow the exact logic proposed in the paper for the proof of concept.

> but would we be interested in different p-norms instead of just l1 difference?
Error is defined as the diff between the original gradient and the one after compression + decompression, and it needs to be applied reversely in the next iteration. Not sure how to do this with p-norm.

> Also, would we be interested in keeping a sort of decaying weighted average (similar to momentum) of errors instead of the error being 100% decided by the previous iteration difference?
Momentum or decaying factor aims to make the learning rate more adaptive as the training progresses and speedup the convergence, but I would be surprised that partially fixing an error with a decaying factor can result in better results. I have thought of making the compression rate adaptive -- compress more aggressively initially, and more conservatively later.

File: torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py (hunk 5 above)
Comment (SciPioneer) [reply torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py:46]: Let's first prove that PowerSGD can be useful as a comm hook, then we can add more tests and documentations later. Currently the speedup looks promising, but the accuracy loss is a concern, so I hope error feedback can bridge the accuracy gap.

File: torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py (hunk 2 above)
Comment (SciPioneer) [reply torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py:55]: Good question! TBH, I don't see the answer in the referenced paper. According to PowerSGD paper, this should be uncommon.

I made the error feedback optional, so the user can choose a low-accuracy PowerSGD compression w/o such error feedback and reduce the peak memory usage.

File: torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py (hunk 4 above)
Comment (SciPioneer) [reply torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py:183]: The error tensor stored in the state will always be updated in place, and there is only a single state instance across iterations.

`input_tensor_cp` is a temporary variable, and the memory will be released automatically after running this hook.

Comment (rohan-varma) [reply torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py:184]: You're right, the update is in place. Thanks for the catch!

File: torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py [hunk 6]
```diff
@@ -141,6 +179,9 @@ def decompress(fut):
         q = fut.value()[0].div_(world_size)
```
[the remark this answers is not in the document]
Comment (tvogels) [reply torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py:179]: Comment on `torch.matmul(matrix.t(), p, out=q)`:
Maybe you can check if everyone still has the same `p` here.

File: torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py [hunk 7]
```diff
@@ -141,6 +179,9 @@ def decompress(fut):
...
         torch.matmul(p, q.t(), out=matrix)
```
Comment (tvogels) [inline torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py:180]: Could this be where the bug is? The `p` here is not from the scope of the `compute_q` callback, but from the parent scope. That tensor is still '*empty*'.

**Review by tvogels — COMMENTED** [review]
@SciPioneer asked me to have a look at this PR. I added a few suggestions for debugging, but in the end I think the bug might be in the use of the tensor `p`:
- in line 160, it is first created `empty`.
- in line 167, a new tensor p is created within the scope of `compute_q`.
- in line 180, the (still empty) global p is used, but it should be the one from line 167.

File: torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py [hunk 8]
```diff
@@ -98,6 +125,17 @@ def powerSGD_hook(
     padded_total_length = square_side_length ** 2
     input_tensor.resize_(padded_total_length)
     input_tensor[total_length:padded_total_length].fill_(0)
+
+    # Incorporate the error from the previous state into the gradients.
+    if state.use_error_feedback:
+        if input_tensor in state.error_dict:
+            input_tensor.add_(state.error_dict[input_tensor])
```
[the remark this answers is not in the document]
Comment (SciPioneer) [reply torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py:132]: This should be fixed it in a follow-up PR: #48757

File: torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (SciPioneer) [reply torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py:179]: > `torch.manual_seed(rng.randint(1_000_000_000))`:

`with torch.random.fork_rng(devices=[]):` pass an empty list to `devices` arg, which means the RNG is not forked on any CUDA device.

[the remark this answers is not in the document]
Comment (SciPioneer) [reply torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py:179]: > A comment on `torch.matmul(matrix, q, out=p)`:
> Did you double-check that all workers have the same `q` here?

Yes

[the remark this answers is not in the document]
Comment (SciPioneer) [reply torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py:179]: > compute_q

Yes, p is used indeed, `p = fut.value()[0]` basically reads the updated value after allreduce(p).

Before` compute_q(fut)`, the following code runs:
1) P = MQ;
2) AllReduce(P)
```
    torch.matmul(matrix, q, out=p)
    allreduce_p_fut = dist.all_reduce(p, group=group_to_use, async_op=True).get_future()
```

[the remark this answers is not in the document]
Comment (SciPioneer) [reply torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py:179]: > Comment on `torch.matmul(matrix.t(), p, out=q)`:
> Maybe you can check if everyone still has the same `p` here.

Can verify that `p` is the same across processes, as it's the result of AllReduce.

File: torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py (hunk 7 above)
Comment (SciPioneer) [reply torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py:180]: The value of `p` in `decompress(fut)` is same as the value in `compute_q ` as well as the parent scope.

The last line `allreduce_p_fut.then(compute_q).then(decompress)` basically chains up the parent scope and the 2 callbacks.

**Comment by SciPioneer** [comment]
> @SciPioneer asked me to have a look at this PR. I added a few suggestions for debugging, but in the end I think the bug might be in the use of the tensor `p`:
> 
> * in line 160, it is first created `empty`.
> * in line 167, a new tensor p is created within the scope of `compute_q`.
> * in line 180, the (still empty) global p is used, but it should be the one from line 167.

See my reply in the same thread:
https://github.com/pytorch/pytorch/pull/48670#discussion_r535632690

It's the same p object throughout the future chain.

**Comment by SciPioneer** [comment]
> At some point, there is a fundamental lower limit to how much you have to communicate.

@tvogels Is there any quick way to roughly estimate such "lower limit"? I guess we can only do tuning by experimentations, and such limit varies by specific cases.

File: torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (tvogels) [reply torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py:179]: > `with torch.random.fork_rng(devices=[]):` pass an empty list to `devices` arg, which means the RNG is not forked on any CUDA device.

Yes, that makes sense. My comment was more about the `torch.manual_seed` which looks like it would seed all devices (even though they haven't been forked).

File: torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py (hunk 7 above)
Comment (tvogels) [reply torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py:180]: Ah, I see. My bad. I misunderstood the future-based all_reduce API. I thought it wouldn't modify its input and rather create a new tensor with the result of averaging.

**Comment by tvogels** [comment]
> @tvogels Is there any quick way to roughly estimate such "lower limit"? I guess we can only do tuning by experimentations, and such limit varies by specific cases.

Good question. I am not aware of any theory-based rule here. I guess you can quickly find it by starting with little compression and repeatedly halving the communication budget until you lose too much accuracy. 

One note: This fundamental limit seems to change during training. I recently saw this [paper](https://arxiv.org/pdf/2010.16248.pdf) that demonstrates that you have to be careful during the beginning of training and when the learning rate decays, but you can get away with stronger compression during the rest of training.

To verify the implementation, maybe you could start with the settings from the [PowerSGD paper](https://arxiv.org/pdf/1905.13727.pdf) (ResNet-18, 16 workers, rank-2 PowerSGD, corresponding to 136× compression) and compare results.
